src/main/resources/assets/carpet/scripts/shapes.scl

```diff
@@ -20,60 +25,90 @@ draw_sphere(args)->(
                 positions += [cx+cyaw,cy+spitch*radius,cz+syaw];
                 positions += [cx+cos(_+180)*cpitch*radius,cy+spitch*radius,cz+sin(_+180)*cpitch*radius],
                 for(range(-syaw,syaw+1),
-                    positions += [cx+cyaw*cpitch,cy+spitch*radius,cz+_]
+                    positions += [cx+cyaw*cpitch,cy+spitch*radius, cz+_]
                 )
             )
         )
     );
     positions
 );
 
-draw_diamond(args)->(
-    [pos, radius] = args;
+//This function generates a hollow diamond, essentially by 3D printing it.
+//It iterates along the radius, and at each point it draws a flat diamond in the y direction
+//Algorithmic efficiency:
+//  O(n^2) cos 2d surface
+draw_diamond(pos, radius)->(
     positions = {};
     c_for(r=0, r<radius, r+=1,
         y = r-radius+1;
         c_for(x=-r,x<=r,x+=1,
             z=r-abs(x);
-            positions += [pos:0+x, pos:1 +y, pos:2+z];
-            positions += [pos:0+x, pos:1 +y, pos:2-z];
-            positions += [pos:0+x, pos:1 -y, pos:2+z];
-            positions += [pos:0+x, pos:1 -y, pos:2-z];
+            positions += [pos:0+x, pos:1+y, pos:2+z];
+            positions += [pos:0+x, pos:1+y, pos:2-z];
+            positions += [pos:0+x, pos:1-y, pos:2+z];
+            positions += [pos:0+x, pos:1-y, pos:2-z];
         )
     );
     positions
 );
 
-draw_filled_diamond(args)->(
-    [pos, radius] = args;
+//Draws a filled diamond by drawing a bunch of hollow diamonds within
+//While the code looks dodgy, it's equivalent to the filling algorithm
+//which fills between the endpoints as opposed to just plotting them
+//Algorithmic efficiency:
+//  O(n^3) cos 3d volume
+draw_filled_diamond(pos, radius)->(
     positions = {};
-    for(diamond(pos,radius,radius),
-        positions += pos(_)
+    loop(radius,
+        positions = positions + draw_diamond(pos, _+1);
     );
     positions
 );
 
-draw_pyramid(args)->(
-    [pos, radius, height, pointing, orientation, fill_type, is_square] = args;//todo change to hollow boolean arg
+//Draws a bunch of circles or squares of decreasing or increasing radii based on where it's pointing.
+//Circles are drawn using _fill_flat_circle
+//Algorithmic efficiency:
+//  O(n^2) for hollow cones / pyramids
+//  O(n^3) for filled cones / pyramids
+draw_pyramid(pos, radius, height, pointing, orientation, fill_type, is_square)->(
```

Comment:
Again no comment on what it does! Ik, ik, the name kinda gives it away, but it's always nice to confirm it more than the name.

Reply:
Same for some others.

Or was this going to have a docs section elsewhere? Can't remember.